Readme.md
Horseman
=========

[![Build Status](https://travis-ci.org/johntitus/node-horseman.svg?branch=master)](https://travis-ci.org/johntitus/node-horseman)

Horseman lets you run [PhantomJS](http://phantomjs.org/) from Node.

Horseman has:

  * a simple chainable (Promise based) API,
  * an easy-to-use control flow (see the examples),
  * support for multiple tabs being open at the same time,
  * built in jQuery for easier page manipulation,
  * built in bluebird for easier in-browser async.

## Examples

### Search on Google

```js
var Horseman = require('node-horseman');
var horseman = new Horseman();

horseman
  .userAgent('Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0')
  .open('http://www.google.com')
  .type('input[name="q"]', 'github')
  .click('[name="btnK"]')
  .keyboardEvent('keypress', 16777221)
  .waitForSelector('div.g')
  .count('div.g')
  .log() // prints out the number of results
  .close();
```

Save the file as `google.js`. Then, `node google.js`.

### Count Twitter Followers Concurrently

```js
const Horseman = require('node-horseman');
const users = ['PhantomJS', 'nodejs'];

users.forEach((user) => {
    const horseman = new Horseman();
    horseman
        .open(`http://twitter.com/${user}`)
        .text('.ProfileNav-item--followers .ProfileNav-value')
        .then((text) => {
            console.log(`${user}: ${text}`);
        })
        .close();
});
```

Save the file as `twitter.js`. Then, `node twitter.js`.

### For longer examples, check out the Examples folder.

## Installation

`npm install node-horseman`

Note: Make sure PhantomJS is available in your path,
you have the phantomjs-prebuilt/phantomjs npm package installed,
or use the `phantomPath` option.

## API

### Setup

#### new Horseman(options)

Create a new instance that can navigate around the web.

The available options are:

  * `timeout`: how long to wait for page loads or wait periods,
    default `5000` ms.
  * `interval`: how frequently to poll for page load state, default `50` ms.
  * `loadImages`: load all inlined images, default `true`.
  * `switchToNewTab`: switch to new tab when created, default `false`.
  * `diskCache`: enable disk cache, default `false`.
  * `diskCachePath`: location for the disk cache. *(requires PhantomJS 2.0.0 or above)*
  * `cookiesFile`: A file where to store/use cookies.
  * `ignoreSSLErrors`: ignores SSL errors,
    such as expired or self-signed certificate errors.
  * `sslProtocol`: sets the SSL protocol for secure connections
    `[sslv3|sslv2|tlsv1|any]`, default `any`.
  * `webSecurity`: enables web security and forbids cross-domain XHR.
  * `injectJquery`: whether jQuery is automatically loaded into each page.
    Default is `true`.
    If jQuery is already present on the page, it is not injected.
  * `jQueryLocation`: the location of the jQuery file to be loaded into the page.
    Defaults to the jQuery file shipped with node-horseman.
  * `injectBluebird`: whether bluebird is automatically loaded into each page.
    Default is `false`.
    If `true` and `Promise` is already present on the page, it is not injected.
    If `'bluebird'` it is always injected as Bluebird,
    whether Promise is present or not.
  * `bluebirdDebug`: whether or not to enable bluebird debug features.
    Default is `false`.
    If `true` non-minified bluebird is injected
    and long stack traces are enabled
  * `proxy`: specify the proxy server to use `address:port`, default not set.
  * `proxyType`: specify the proxy server type `[http|socks5|none]`,
    default not set.
  * `proxyAuth`: specify the auth information for the proxy `user:pass`,
    default not set.
  * `phantomPath`: If PhantomJS is not installed in your path,
    you can use this option to specify the executable's location.
  * `phantomOptions`: Explicit PhantomJS options, e.g.
    `{'ssl-certificates-path': 'ca.pem'}`.
    For a complete list refer to the [PhantomJS command line interface](
    http://phantomjs.org/api/command-line.html).
    **These options have precedence over options implicitly set by Horseman.**
  * `debugPort`: Enable web inspector on specified port, default not set.
  * `debugAutorun`: Autorun on launch when in debug mode, default is true.

### Configuration

#### .setProxy(ip, \[port\], \[type\], \[user, pass\])

Dynamically set proxy settings (***requires PhantomJS 2.0.0 or above***).
The `ip` argument can either be the IP of the proxy server,
or a URI of the form `type://user:pass@ip:port`.

The `port` is optional and defaults to `80`.
The `type` is optional and defaults to `'http'`.
The `user` and `pass` are the optional username and password for authentication,
by default no authentication is used.

### Cleanup

Be sure to `.close()` each Horseman instance when you're done with it!

#### .close()

Closes the Horseman instance by shutting down PhantomJS.

### Navigation

#### .open(url)

Load the page at `url`.

#### .post(url, postData)

POST `postData` to the page at `url`.

#### .put(url, putData)

PUT `putData` to the page at `url`.

#### .back()

Go back to the previous page.

#### .forward()

Go forward to the next page.

#### .status()

The HTTP status code returned for the page just opened.

#### .reload()

Refresh the current page.

#### .cookies(\[object|array of objects|string\])

Without any options,
this function will return all the cookies inside the browser.

```js
horseman
  .open('http://httpbin.org/cookies')
  .cookies()
  .log() // []
  .close();
```

You can pass in a cookie object to add to the cookie jar.

```js
horseman
  .cookies({
    name: 'test',
    value: 'cookie',
    domain: 'httpbin.org'
  })
  .open('http://httpbin.org/cookies')
  .cookies()
  .then(function(cookies){
    console.log(cookies);
    return horseman.close();
  });

/*
[ { domain: '.httpbin.org',
    httponly: false,
    name: 'test',
    path: '/',
    secure: false,
    value: 'cookie' } ]
*/
```

You can pass in an array of cookie objects
to reset all the cookies in the cookie jar
(or pass an empty array to remove all cookies).

```js
horseman
  .cookies([
  {
    name : 'test2',
    value : 'cookie2',
    domain: 'httpbin.org'
  },
  {
    name : 'test3',
    value : 'cookie3',
    domain: 'httpbin.org'
  }])
  .open('http://httpbin.org/cookies')
  .cookies()
  .then(function(cookies){
    console.log(cookies.length); // 2
    return horseman.close();
  });

```

[cookies.txt]: <http://www.cookiecentral.com/faq/#3.5>
You can pass in the name of a [cookies.txt][] formatted file
to reset all the cookies in the cookie jar
to those contained in the file.

```js
horseman
  .cookies('my-cookies.txt')
  .open('http://httpbin.org/cookies')
  .cookies()
  .then(function(cookies){
    console.log(cookies);
    return horseman.close();
  });

  /* Cookies from my-cookies.txt (converted to the above object format) */

```

#### .userAgent(userAgent)

Set the `userAgent` used by PhantomJS.
You have to set the userAgent before calling `.open()`.

#### .headers(headers)

Set the `headers` used when requesting a page.
The headers are a javascript object.
You have to set the headers before calling `.open()`.

#### .authentication(user, password)

Set the `user` and `password` for accessing a web page
using basic authentication.
Be sure to set it before calling `.open(url)`.

```js
horseman
  .authentication('myUserName', 'myPassword')
  .open('http://httpbin.org/basic-auth/myUserName/myPassword')
  .html('pre')
  .then(function(body) {
    console.log(body);
    /*
        {
        "authenticated": true,
        "user": "myUserName"
      }
    */
    return horseman.close();
  });
```

#### .viewport(width, height)

Set the `width` and `height` of the viewport, useful for screenshotting.
You have to set the viewport before calling `.open()`.

#### .scrollTo(top, left)

Scroll to a position on the page,
relative to the top left corner of the document.

#### .zoom(zoomFactor)

Set the amount of zoom on a page.  The default zoomFactor is 1.
To zoom to 200%, use a zoomFactor of 2.
Combine this with `viewport` to produce high DPI screenshots.

```js
horseman
  .viewport(3200,1800)
  .zoom(2)
  .open('http://www.horsemanjs.org')
  .screenshot('big.png')
  .close();
```

### Evaluation

Evaluation elements return information from the page.

#### .title()

Get the title of the current page.

#### .url()

Get the URL of the current page.

#### .visible(selector)

Determines if a selector is visible, or not, on the page. Returns a boolean.

#### .exists(selector)

Determines if the selector exists, or not, on the page. Returns a boolean.

#### .count(selector)

Counts the number of `selector` on the page. Returns a number.

#### .html(\[selector\], \[file\])

Gets the HTML inside of an element.
If no `selector` is provided, it returns the HTML of the entire page.
If `file` is provided, the HTML will be written to that filename.

#### .text(selector)

Gets the text inside of an element.

#### .plainText()

Gets the plain text of the whole page (using PhantomJS's [`plainText`](http://phantomjs.org/api/webpage/property/plain-text.html) property).

#### .value(selector, \[val\])

Get, or set, the value of an element.

#### .attribute(selector, attribute)

Gets an attribute of an element.

#### .cssProperty(selector, property)

Gets a CSS property of an element.

#### .width(selector)

Gets the width of an element.

#### .height(selector)

Gets the height of an element.

#### .screenshot(path)

Saves a screenshot of the current page to the specified `path`.
Useful for debugging.

#### .screenshotBase64(type)

Returns a base64 encoded string representing the screenshot.
Type must be one of 'PNG', 'GIF', or 'JPEG'.

#### .crop(area, path)

[getBoundingClientRect]: https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect
Takes a cropped screenshot of the page.
`area` can be a string identifying an html element on the screen to crop to,
or a [getBoundingClientRect][] object.

#### .cropBase64(area, type)

Returns a string representing a cropped, base64 encoded screenshot of the page.
`area` can be a string identifying an html element on the screen to crop to,
or a [getBoundingClientRect][] object.
Type must be one of 'PNG', 'GIF', or 'JPEG'.

#### .pdf(path, \[paperSize\])

[US Letter]: <http://en.wikipedia.org/wiki/Letter_%28paper_size%29>
Renders the page as a PDF.
The default paperSize is [US Letter][].

The `paperSize` object should be in either this format:

```js
{
  width: '200px',
  height: '300px',
  margin: '0px'
}
```

or this format

```js
{
  format: 'A4',
  orientation: 'portrait',
  margin: '1cm'
}
```

Supported formats are: `A3`, `A4`, `A5`, `Legal`, `Letter`, `Tabloid`.

Orientation (`portrait`, `landscape`) is optional and defaults to 'portrait'.

Supported dimension units are: 'mm', 'cm', 'in', 'px'. No unit means 'px'.

You can create a header and footer like this:

```js
horseman
  .open('http://www.amazon.com')
  .pdf('amazon.pdf', {
    format: 'Letter',
    orientation: 'portrait',
    margin: '0.5in',
    header: {
      height: '3cm',
      contents: function(pageNum, numPages) {
        if (pageNum == 1) {
          return '';
        }
        return '<h3>Header ' + pageNum + ' / ' + numPages + '</h3>';
      }
    },
    footer: {
      height: '3cm',
      contents: function(pageNum, numPages) {
        if (pageNum == 1) {
          return '';
        }
        return '<h3>Footer ' + pageNum + ' / ' + numPages + '</h3>';
      }
    }
  })
  .close()
```

#### .log()

Outputs the results of the last call in the chain, or a string you provide,
without breaking the chain.

```js
horseman
  .open('http://www.google.com')
  .count('a')
  .log() // outputs the number of anchor tags
  .click('a')
  .log('clicked the button') //outputs the string
  .close();
```

#### .do(fn)

Run an function without breaking the chain. Works with asynchronous functions.
Must call the callback when complete.

```js
horseman
  .open('http://www.google.com')
  .do(function(done){
    setTimeout(done,1000);
  })
  .close();
```

#### .evaluate(fn, \[arg1, arg2,...\])

Invokes `fn` on the page with args. On completion it returns a value.
Useful for extracting information from the page.

```js
horseman
  .open('http://en.wikipedia.org/wiki/Headless_Horseman')
  .evaluate( function(selector){
      // This code is executed inside the browser.
      // It's sandboxed from Node, and has no access to anything
      // in Node scope, unless you pass it in, like we did with 'selector'.
      //
      // You do have access to jQuery, via $, automatically.
      return {
        height : $( selector ).height(),
        width : $( selector ).width()
      }
    }, '.thumbimage')
  .then(function(size){
    console.log(size);
    return horseman.close();
  });
```

Can be used in an asynchronous way as well (with a node-style callback).
This is similar to `.do`, but `fn` is invoked in the browser.

```js
horseman
  .open('http://en.wikipedia.org/wiki/Headless_Horseman')
  .evaluate(function(ms, done){
    var start = Date.now();
    setTimeout(function() {
      done(null, Date.now() - start);
      // ^ Can pass an Error as first argument,
      // making evaluate action reject its Promise in Node.
      // Second argument is what the Promise will resolve to.
    }, ms);
  }, 100)
  .then(function(actualMs){
    console.log(actualMs);
  })
  .close();
```

Lastly, if `fn` returns a Promise or thenable,
it will be waited on and the action in Node will resolve/reject accordingly.

```js
horseman
  .open('http://en.wikipedia.org/wiki/Headless_Horseman')
  .evaluate(function() {
    // Silly example for illustrative purposes.
    return Bluebird.delay(100).return('Hello World');
  })
  .then(function(mesg){
    // Will log 'Hello World' after a roughly 100 ms delay.
    console.log(mesg);
  })
  .close();
```

#### .click(selector)

Clicks the `selector` element once.

#### .select(selector, value)

Sets the value of a `select` element to `value`.

#### .clear(selector)

Sets the value of an element to `''`.

#### .type(selector, text, \[options\])

Enters the `text` provided into the `selector` element.
Options is an object containing `eventType`
(`'keypress'`, `'keyup'`, `'keydown'`. Default is `'keypress'`)
and `modifiers`, which is a string in the form of `ctrl+shift+alt`.

#### .upload(selector, path)

Specify the `path` to upload into a file input `selector` element.

#### .download(url, \[path\], \[binary\])

Download the contents of `url`.
If `path` is supplied the contents will be written there,
otherwise this gets the contents.
If `binary` is `true` it gets the contents as a node `Buffer`,
otherwise it gets them as a string (`binary` defaults to `false`).

***Please note: binary downloads do not work correctly with PhantomJS 1.x***


#### .injectJs(file)

Inject a JavaScript file onto the page.

#### .includeJs(url)

Include an external JavaScript script on the page via URL.

#### .mouseEvent(type, \[x, y, \[button\]\])

Send a mouse event to the page.
Each event is sent to the page as if it comes from real user interaction.
`type` must be one of
`'mouseup'`, `'mousedown'`, `'mousemove'`, `'doubleclick'`, or `'click'`,
which is the default.
`x` and `y` are optional
and specify the location on the page to send the mouse event.
`button` is also optional, and defaults to `'left'`.

#### .keyboardEvent(type, key, \[modifier\])

[phantom keys]: <https://github.com/ariya/phantomjs/commit/cab2635e66d74b7e665c44400b8b20a8f225153a>
Send a keyboard event to the page.
Each event is sent to the page as if it comes from real user interaction.
`type` must be one of `'keyup'`, `'keydown'`, or `'keypress'`,
which is the default.
`key` should be a numerical value from [this page][phantom keys].
For instance, to send an "enter" key press,
use `.keyboardEvent('keypress',16777221)`.

`modifier` is optional, and comes from this list:

  * 0x02000000: A Shift key on the keyboard is pressed
  * 0x04000000: A Ctrl key on the keyboard is pressed
  * 0x08000000: An Alt key on the keyboard is pressed
  * 0x10000000: A Meta key on the keyboard is pressed
  * 0x20000000: A keypad button is pressed

To send a shift+p event,
you would use `.keyboardEvent('keypress','p',0x02000000)`.

### Waiting

These functions for the browser to wait for an event to occur.
If the event does not occur before the timeout period
(configurable via the options),
a timeout event will be fired and the Promise for the action will reject.

#### .wait(ms)

Wait for `ms` milliseconds e.g. `.wait(5000)`

#### .waitForNextPage([options])

Wait until a page finishes loading, typically after a `.click()`.
`options` can have be `{timeout: 5000}` to define an action-specific timeout.

#### .waitForSelector(selector, [options])

Wait until the element `selector` is present,
e.g., `.waitForSelector('#pay-button')`

`options` can have be `{timeout: 5000}` to define an action-specific timeout.

#### .waitFor(fn, \[arg1, arg2,...\], value)

Wait until the `fn` evaluated on the page returns the *specified* `value`.
`fn` is invoked with args.

```js
// This will call the function in the browser repeatedly
// until true (or whatever else you specified) is returned
horseman
  .waitFor(function waitForSelectorCount(selector, count) {
    return $(selector).length >= count

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 3.3.0 - 2017-04-06
### Added
- `phantomOptions` option for passing arbitrary PhantomJS CLI options -
  thanks @fabiocorneti
- Support for per-action timeouts on `.waitFor` type actions - thanks @piercus

### Fixed
- `.boundingRectangle()` sometimes not using jQuery when it could -
  thanks @robsgreen
- `.evaluate()` sometimes failing due to an utilized internal variable -
  thanks @robertpallas

### Changed
- Removed `clientScripts` option (it had no effect anyways)

## 3.2.0 - 2016-10-21
### Added
- `diskCache` and `diskCachePath` options - thanks @efernandesng

## 3.1.1 - 2016-06-17
Republish 3.1.0 without unwanted files which broke Windows installation

## 3.1.0 - 2016-06-15
### Added
- .plainText() action - thanks @mzhangARS

### Changed
- Removed `port` constructor option (it no longer had any effect)

## 3.0.1 - 2016-04-12
### Fixed
- .injectJs() now properly rejects when it fails
- #158 - bug with `injectBluebird` option

## 3.0.0 - 2016-04-06
### Added
- Support for adding custom actions via `Horseman.registerAction`

## 3.0.0-beta2 - 2016-04-05
### Added
- .closeTab() action removes a tab and closes its page
- tabClosed event fired when a tab is closed
  (by .closeTab() or something else)
- .frameName() action
- .frameCount() action
- .frameNames() action
- .switchToFocusedFrame() action
- .switchToFrame() action
- .switchToMainFrame() action
- .switchToParentFrame() action
- .download() action allows getting the raw contents of a URL
  (even works for binary files is using PhantomJS 2 or newer)

### Changed
- .log() now preserves resolution value of previous action
- More output added to verbose debugging

### Fixed
- More readable stack traces from .evaluate() rejections
  (especially with long stack traces on)
- Occasional bug with returning undefined in .evaluate() function

### Deprecated
- .switchToChildFrame() action is deprecated
  since PhantomJS deprecated it

## 3.0.0-beta1 - 2016-03-22
### Added
- bluebird 3 functions are now chainable with Horseman actions
- Support for callback/Promises in .evaluate() functions
- .at() action allows registering callbacks to run in PhantomJS JavaScript
  environment (see Reame for details)
- .setProxy() action allows changing Phantom's proxy settings
  (requires PhantomJS 2.0.0 or newer)
- .includeJs() action allows injecting JavaScript from a URL
- .put() action
- Support for SlimerJS as well as PhantomJS - thanks @w33ble
- Support for PhantomJS remote debugger - thanks @dustinblackman
- Support for reading cookies.txt files
- Options for injecting bluebird into each page for using Promises in browser
- Option for automatically switching to newly opened tabs
- Option for more verbose debug

### Changed
- Horseman Promises now reject on error
- Waiting actions now reject on timeout (timeout callback is still called too)
- Phantom's reourceTimeout now set to match Horseman's timeout
- Horseman will use Phantom from the phantomjs-prebuilt/phantomjs npm packages
  if installed (still overridden by phantomPath option)

### Fixed
- .open() now rejects on fail - thanks @w33ble.
- .crop() now works with zoomFactor other than 1 - thanks @wong2.
- .waitForSelector() now uses JQuery for selectors (when avialable)
- navigationRequested callback now called with correct parameters -
  thanks @grahamkennery.
- Tabs/pages not opened by Horseman (e.g. by a link)
  are now correctly added to Horseman's list of tabs

### Updated
- bluebird dependency to 3.0.1
- node-phantom-simple dependency to 2.2.4
- clone dependency to 1.0.2

## 2.8.2 - 2015-11-23
### Fixed
- #74 - do not try to exit PhantomJS when it failed to initialize

## 2.8.1 - 2015-11-06
### Fixed
- .upload() now works in PhantomJS 1.x line (still broken in 2.x line) -
  thanks @flashhhh
- logging for falsy arguments - thanks @awlayton

## 2.8.0 - 2015-10-28
### Added
- .cropBase64(). Thanks @jeprojects!

### Updated
- node-phantom-simple dependency to version 2.1.0.
- Updated readme with examples on how to set header/footer contents.

## 2.7.1 - 2015-10-21
### Fixed
- If an exception was thrown, close() wouldn't get called. Merged #58.
  Thanks @awlayton!

## 2.7.0 - 2015-10-15
### Updated
- Makes promise functions (like then, catch, tap) chainable with the rest of
  Horseman actions. Merges #57. Thanks @awlayton!

## 2.6.0 - 2015-10-08
### Updated
- .close() is now chainable. Updated readme with examples.

## 2.5.0 - 2015-10-08
### Added
- .log(). Closes #54.

### Fixed
- .viewport() wasn't returning the actual viewport size.

## 2.4.0 - 2015-10-08
### Added
- .do(). Closes #53.

## 2.3.0 - 2015-10-01
### Added
- Support for multiple tabs. Closes #49.
- .status(). Returns the HTTP status code from the last opened page.

## 2.2.1 - 2015-10-01
### Fixed
- `.waitFor` bug fix. Closes #45.

## 2.2.0 - 2015-09-29
### Added
- `.crop()`. Closes #51.

## 2.1.0 - 2015-09-27
### Added
- API is chainable again, thanks to @awlayton. Addresses #46.

## 2.0.2 - 2015-09-21
### Fixed
- `waitForNextPage` was broken. Fixes #44 (thanks @edge)

## 2.0.1 - 2015-09-14
### Upgraded
- Moved to `node-phantom-simple` 2.0.4.

## 2.0.0 - 2015-09-10
### Changed
- Complete API rewrite to use Promises.
- Removed dependency on `deasync`.
- Tab support removed, will add back in the future.

## 1.5.6 - 2015-08-17
### Fixed
- #39. Upgraded to node-phantom-simple 2.0.3.

## 1.5.5 - 2015-08-12
### Changed
- Upgraded to node-phantom-simple 2.0.2. Close #37.
- Upgraded to Mocha 2.2.5 to remove "child_process: customFds option is
  deprecated, use stdio instead" message when testing.
- Edited this document to reflect actual release dates from 1.4.1 - present.

## 1.5.4 - 2015-07-22
### Fixed
- Merges #36 - Setting via horseman.value( newVal ) now fires a change event.
  Thanks @fpinzn.

## 1.5.3 - 2015-07-13
### Fixed
- Fix #27 & 33 - .cookies() wasn't returning a list of cookies for the current
  page. Now fixed.

## 1.5.2 - 2015-07-01

lib/index.js
'use strict';

var clone = require('clone');
var defaults = require('defaults');
var phantom = require('node-phantom-simple');
var path = require('path');
var debug = require('debug')('horseman');
var debugv = require('debug')('horseman:verbose');
var actions = require('./actions');
var HorsemanPromise = require('./HorsemanPromise.js');
var noop = function() {};

/**
 * Check for npm PhantomJS
 */
var phantomjs;
try {
	phantomjs = require('phantomjs-prebuilt');
	debug('using PhantomJS from phantomjs-prebuilt module');
} catch (err) {
	try {
		phantomjs = require('phantomjs');
		debug('using PhantomJS from phantomjs module');
	} catch (err) {
		phantomjs = {};
		debug('using PhantomJS from $PATH');
	}
}

/**
 * Default options.
 */
var DEFAULTS = {
	timeout: 5000,
	interval: 50,
	weak: true,
	loadImages: true,
	sslProtocol: 'any', //sslv3, sslv2, tlsv1, any
	injectJquery: true,
	switchToNewTab: false,
	injectBluebird: false,
	phantomPath: phantomjs.path,
	jQueryLocation: path.join(__dirname, '../files/jquery-2.1.1.min.js')
};

/**
 * Give each instance a unique ID for debug purposes
 */
var instanceId = 0;



function prepare(horseman, instantiationOptions) {
	return new HorsemanPromise(function(resolve, reject) {
		phantom.create(instantiationOptions, function(err, instance) {
			debug('phantom created');
			horseman.phantom = instance;
			if (err) {
				return reject(err);
			}
			if (debugv.enabled) {
				horseman.phantom.onError = function(msg, trace) {
					var str = msg;
					trace.forEach(function(t) {
						var fun = t.function || '<anonymous>';
						str += '\n\t' + t.file + ': ' + t.line + ' in ' + fun;
					});
					return debugv('onPhantomError', str, horseman.id);
				};
			}
			instance.get('version', function(err, ver) {
				if (err) { return reject(err); }
				var version = ver.major + '.' + ver.minor + '.' + ver.patch;
				debug('phantom version ' + version);
				return horseman.pageMaker()
					.then(resolve)
					.catch(reject);
			});
		});
	})
	.then(function() {
		horseman.pageCnt = 0;
	})
	.bind(horseman);
}

/**
 * Creates a new Horseman.
 * @constructor
 */
var Horseman = function Horseman(options) {
	this.ready = false;
	if (!(this instanceof Horseman)) { return new Horseman(options); }
	this.options = defaults(clone(options) || {}, DEFAULTS);

	this.id = ++instanceId;
	debug('.setup() creating phantom instance %s', this.id);

	var phantomOptions = {
		'load-images': this.options.loadImages,
		'ssl-protocol': this.options.sslProtocol
	};

	if (typeof this.options.ignoreSSLErrors !== 'undefined') {
		phantomOptions['ignore-ssl-errors'] = this.options.ignoreSSLErrors;
	}
	if (typeof this.options.webSecurity !== 'undefined') {
		phantomOptions['web-security'] = this.options.webSecurity;
	}
	if (typeof this.options.proxy !== 'undefined') {
		phantomOptions.proxy = this.options.proxy;
	}
	if (typeof this.options.proxyType !== 'undefined') {
		phantomOptions['proxy-type'] = this.options.proxyType;
	}
	if (typeof this.options.proxyAuth !== 'undefined') {
		phantomOptions['proxy-auth'] = this.options.proxyAuth;
	}
	if (typeof this.options.diskCache !== 'undefined') {
		phantomOptions['disk-cache'] = this.options.diskCache;
	}
	if (typeof this.options.diskCachePath !== 'undefined') {
		phantomOptions['disk-cache-path'] = this.options.diskCachePath;
	}
	if (typeof this.options.cookiesFile !== 'undefined') {
		phantomOptions['cookies-file'] = this.options.cookiesFile;
	}

	if (this.options.debugPort) {
		phantomOptions['remote-debugger-port'] = this.options.debugPort;
		phantomOptions['remote-debugger-autorun'] = 'no';
		if (this.options.debugAutorun !== false) {
			phantomOptions['remote-debugger-autorun'] = 'yes';
		}
	}

	Object.keys(this.options.phantomOptions || {}).forEach(function (key) {
		if (typeof phantomOptions[key] !== 'undefined') {
			debug('Horseman option ' + key + ' overridden by phantomOptions');
		}
		phantomOptions[key] = this.options.phantomOptions[key];
	}.bind(this));

	var instantiationOptions = {
		parameters: phantomOptions
	};

	if (typeof this.options.phantomPath !== 'undefined') {
		instantiationOptions['path'] = this.options.phantomPath;
	}

	// Store the url that was requested for the current url
	this.targetUrl = null;

	// Store the HTTP status code for resources requested.
	this.responses = {};

	this.tabs = [];
	this.onTabCreated = noop;
	this.onTabClosed = noop;

	this.ready = prepare(this, instantiationOptions);
};

/**
 * Attaches a page to the Horseman instance.
 * @param {string} [url=about:blank] - URL to open.
 * @param {Page} [_page] - Page to attach. If null, a new Page will be created.
 * @emits Horseman#tabCreated
 */
Horseman.prototype.pageMaker = function(url, _page) {
	var self = this;
	url = url || 'about:blank';

	// .try runs synchronously
	var loaded;
	return HorsemanPromise.try(function getPage() {
			var page;
			if (_page) {
				var p = setupPage(_page);
				loaded = _page.loadedPromise();
				return p;
			} else {
				return HorsemanPromise.fromCallback(function(done) {
						return self.phantom.createPage(function(err, page) {
							if (page) {
								var p = setupPage(page).asCallback(done);
								loaded = page.loadedPromise();
								return p;
							}
							return done(err, page);
						});
					})
					.tap(function openPage(page) {
						return HorsemanPromise.fromCallback(function(done) {
							return page.open(url, done);
						});
					});
			}
		})
		.tap(function waitForLoadFinished() {
			return loaded;
		})
		.tap(function attachPage(page) {
			if (!_page || self.options.switchToNewTab) {
				self.page = page;
			}
			var tabNum = self.tabs.push(page) - 1;
			return self.onTabCreated(tabNum, page);
		})
		.bind(self);

	/**
	 * Needs to be called *BEFORE* page creation callback returns
	 * This is necessary to avoid having events before callbacks are attached
	 * @param page - Newly created page to decorate and attach to horseman
	 */
	function setupPage(page) {
		debug('page created');

		page.onPageCreated = function(newPage) {
			return self.ready = self.pageMaker(undefined, newPage)
				.tap(function() {
					page.onPageCreated2(newPage);
				});
		};
		page.onPageCreated2 = noop;

		page.onResourceReceived = function(response) {
			self.responses[response.url] = response.status;
		};

		page.onLoadFinished = loadFinishedSetup;

		page.onClosing = function() {
			var tabNum = self.tabs.indexOf(page);
			debug('closing tab', tabNum);
			self.tabs.splice(tabNum, 1);
			if (self.page === page) {
				// Switch to previous tab when current tab closes
				self.page = self.tabs[tabNum - 1];
			}
			self.onTabClosed(tabNum, page);
		};

		page.onTimeout = noop;

		// Create a Promise for when onLoadFinished callback is done
		page.loadedPromise = function() {
			return HorsemanPromise.fromCallback(function(done) {
				page.loadDone = function(err, res) {
					page.loadDone = undefined;
					return done(err, res);
				};
			});
		};

		// If verbose debug, add default error and consoleMessage handlers
		if (debugv.enabled) {
			page.onError = function(msg, trace) {
				var str = msg;
				trace.forEach(function(t) {
					var fun = t.function || '<anonymous>';
					str += '\n\t' + t.file + ': ' + t.line + ' in ' + fun;
				});
				return debugv('onError', str, self.id);
			};
			page.onConsoleMessage = function(msg, lineNum, sourceId) {
				var str = msg + ' line: ' + lineNum + ' in ' + sourceId;
				return debugv('onConsoleMessage', str, self.id);
			};
		}

		return HorsemanPromise.fromCallback(function(done) {
			return page.set(
				'settings.resourceTimeout',
				self.options.timeout,
				done
			);
		}).return(page);

		/**
		 * Do any javascript injection on the page
		 * TODO: Consolidate into one page.evaluate?
		 */
		function loadFinishedSetup(status) {
			var args = arguments;
			self.pageCnt++;
			debug('phantomjs onLoadFinished triggered', status, self.pageCnt);

			return self.ready = HorsemanPromise.try(function checkStatus() {
				if (status !== 'success') {
					var err = new Error('Failed to load url');
					return HorsemanPromise.reject(err);
				}
			})
			.then(function injectJQuery() {
				if (!self.options.injectJquery) {
					return;
				}

				return HorsemanPromise.fromCallback(function hasJQuery(done) {
					return page.evaluate(function hasJQuery() {
						return (typeof window.jQuery !== 'undefined');
					}, done);
				})
				.then(function(hasJquery) {
					if (hasJquery) {
						debug('jQuery not injected - already exists on page');
						return;
					}

					return HorsemanPromise.fromCallback(function(done) {
						return page.injectJs(self.options.jQueryLocation, done);
					})
					.tap(function(successful) {
						if (!successful) {
							var err = new Error('jQuery injection failed');
							return HorsemanPromise.reject(err);
						}
						debug('injected jQuery');
					});
				});
			})
			.then(function injectBluebird() {
				var inject = self.options.injectBluebird;
				if (!inject) {
					return;
				}

				return HorsemanPromise.fromCallback(function hasPromise(done) {
					return page.evaluate(function hasPromise() {
						return (typeof window.Promise !== 'undefined');
					}, done);
				})
				.then(function(hasPromise) {
					if (hasPromise && inject !== 'bluebird') {
						debug('bluebird not injected - ' +
								'Promise already exists on page');
						return;
					}

					var bbLoc = 'bluebird/js/browser/bluebird' +
							(self.options.bluebirdDebug ? '' : '.min') + '.js';
					return HorsemanPromise.fromCallback(function(done) {
						return page.injectJs(require.resolve(bbLoc), done);
					})
					.tap(function(successful) {
						if (!successful) {
							var err = new Error('bluebird injection failed');
							return HorsemanPromise.reject(err);
						}
						debug('injected bluebird');
					});

				})
				.then(function configBluebird() {
					return HorsemanPromise.fromCallback(function(done) {
						return page.evaluate(
							function configBluebird(noConflict, debug) {
								if (debug) {
									// TODO: Turn on warnings in bluebird 3
									Promise.longStackTraces();
								}
								if (noConflict) {
									window.Bluebird = Promise.noConflict();
								}
							},
							inject === 'bluebird',
							self.options.bluebirdDebug,
						done);
					});
				});
			})
			.then(function initWindow() {
				return HorsemanPromise.fromCallback(function(done) {
					return page.evaluate(function initWindow() {
						window.__horseman = {};
					}, done);
				});
			})
			.then(function finishLoad() {
				if (page.onLoadFinished2) {
					return page.onLoadFinished2.apply(page, args);
				}
			})
			.bind(self)
			.asCallback(page.loadDone);
		}
	}
};

//**************************************************//
// Actions
//**************************************************//
/**
 * Add an action to the Horseman (and HorsemanPromise) prototype.
 * @param {string} name - Name to access action with
 * @param {function} action - Code for the action.
 * this will be bound to the horseman instance.
 */
Horseman.registerAction = function registerAction(name, action) {
	if (typeof name === 'function') {
		// Work with just a named function
		action = name;
		name = action.name;
	}

	// Keep track of page counter before each action
	Horseman.prototype[name] = function() {
		this.lastActionPageCnt = this.curActionPageCnt;
		this.curActionPageCnt = this.pageCnt;
		return action.apply(this, arguments);
	};

	// Allow chaining actions off HorsemanPromises
	HorsemanPromise.prototype[name] = function() {
		var args = arguments;
		return this.then(function(val) {
			this.lastVal = val;
			return this[name].apply(this, args);
		});
	};
};

/**
 * Attach all of the actions.
 */
Object.keys(actions).forEach(function attachAction(name) {
	Horseman.registerAction(name, actions[name]);
});

//**************************************************//
// Cleanup
//**************************************************//
Horseman.prototype.close = function() {
	debug('.close().');
	var self = this;
	return this.ready.finally(function() {
		var p;
		if (!self.closed && self.phantom) {
			p = HorsemanPromise.fromCallback(function(done) {
				return self.phantom.exit(done);
			});
		}
		//self.ready = HorsemanPromise.reject(new Error('phantom is closed'));
		self.closed = true;
		return p || HorsemanPromise.resolve();
	});
};
HorsemanPromise.prototype.close = function() {
	var self = this;
	return this
		.finally(function() {
			return this.close();
		}).finally(function() {
			return self;
		});
};

module.exports = Horseman;
module.exports.TimeoutError = HorsemanPromise.TimeoutError;
